### PR TITLE
Add build init support for C++ applications and libraries

### PR DESCRIPTION
**Pull request by ghale**
Adds support for generating C++ applications and libraries with the "init" command.

### CODE REVIEW

File: subprojects/build-init/src/main/resources/org/gradle/buildinit/tasks/templates/cppapp/app.h.template [hunk 1]
```diff
@@ -0,0 +1,11 @@
+/*
```
Comment (big-guy) [inline subprojects/build-init/src/main/resources/org/gradle/buildinit/tasks/templates/cppapp/app.h.template]: This needs header guards or pragma once

File: subprojects/build-init/src/main/resources/org/gradle/buildinit/tasks/templates/cpplibrary/hello.h.template [hunk 2]
```diff
@@ -0,0 +1,18 @@
+/*
```
Comment (big-guy) [inline subprojects/build-init/src/main/resources/org/gradle/buildinit/tasks/templates/cpplibrary/hello.h.template]: Needs header guards/pragma once

File: subprojects/build-init/src/main/java/org/gradle/buildinit/tasks/InitBuild.java [hunk 3]
```diff
@@ -272,7 +272,7 @@ public void setProjectName(String projectName) {
      * @since 5.0
      */
     @Incubating
-    @Option(option = "package", description = "Set the package for source files.")
+    @Option(option = "package", description = "Set the package for source files when generating JVM projects or the namespace for source files when generating C++ projects.")
```
Comment (big-guy) [inline subprojects/build-init/src/main/java/org/gradle/buildinit/tasks/InitBuild.java]: +1 for adam's suggestion of ignoring this for C++ for now

File: subprojects/build-init/src/main/resources/org/gradle/buildinit/tasks/templates/cpplibrary/hello.h.template [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+ * This C++ source file was generated by the Gradle 'init' task.
+ */
+
+#ifdef _WIN32
+#define EXPORT_FUNC __declspec(dllexport)
```
[the remark this answers is not in the document]
Comment (ghale) [reply subprojects/build-init/src/main/resources/org/gradle/buildinit/tasks/templates/cpplibrary/hello.h.template]: Done

File: subprojects/build-init/src/main/java/org/gradle/buildinit/plugins/internal/CppProjectInitDescriptor.java [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+public abstract class CppProjectInitDescriptor extends LanguageLibraryProjectInitDescriptor {
+    public CppProjectInitDescriptor(BuildScriptBuilderFactory scriptBuilderFactory, TemplateOperationFactory templateOperationFactory, FileResolver fileResolver, TemplateLibraryVersionProvider libraryVersionProvider) {
+        super("cpp", scriptBuilderFactory, templateOperationFactory, fileResolver, libraryVersionProvider);
+    }
+
+    @Override
+    protected void generate(InitSettings settings, BuildScriptBuilder buildScriptBuilder) {
+        buildScriptBuilder
+            .fileComment("This generated file contains a sample CPP project to get you started.");
+        configureBuildScript(settings, buildScriptBuilder);
```
[the remark this answers is not in the document]
Comment (ghale) [reply subprojects/build-init/src/main/java/org/gradle/buildinit/plugins/internal/CppProjectInitDescriptor.java]: Yeah, I went back and forth with that - I've added it.

File: subprojects/build-init/src/main/resources/org/gradle/buildinit/tasks/templates/cppapp/app.h.template (hunk 1 above)
Comment (ghale) [reply subprojects/build-init/src/main/resources/org/gradle/buildinit/tasks/templates/cppapp/app.h.template]: Done.

File: subprojects/build-init/src/main/resources/org/gradle/buildinit/tasks/templates/cpplibrary/hello.h.template (hunk 2 above)
Comment (ghale) [reply subprojects/build-init/src/main/resources/org/gradle/buildinit/tasks/templates/cpplibrary/hello.h.template]: Done.

File: subprojects/build-init/src/main/java/org/gradle/buildinit/tasks/InitBuild.java (hunk 3 above)
Comment (ghale) [reply subprojects/build-init/src/main/java/org/gradle/buildinit/tasks/InitBuild.java]: It now defaults to the project name.

**Pull request merged**